runners/flink/src/test/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapperTest.java

```diff
@@ -626,13 +650,20 @@ public void testSourceWithNoReaderDoesNotShutdown() throws Exception {
 
       try {
         thread.start();
-        List<UnboundedSource.UnboundedReader<KV<Integer, Integer>>> localReaders =
-            sourceWrapper.getLocalReaders();
-        while (localReaders != null && !localReaders.isEmpty()) {
+        // Wait to see if the wrapper shuts down immediately in case it doesn't have readers
```

Comment:
Ideally we would check for the desired state instead of the assumption that isRunning would flip within 200ms. But short of checking that `finalizeSource` is in the run threads call stack I cannot see any way of doing that.

Reply:
Thank you. Good improvement. Merged to this PR. It still doesn't give 100% confidence because the finalize method will always be hit, but we would see a test failure if it exited the method.

Reply:
Right, but the check is for sleep && finalizeSource, which will only occur when the source is not shutting down.